Added VLMTransformer to do PHI detection using VLMs

yenlow opened the pull request:
Added VLMTransformer to do PHI detection using VLMs, wrapped with pandas udf

TODO:
1. Add redaction by EasyOCR

dmoore247 commented on notebooks/DE-ID/DICOM - VLMTransformer.ipynb:
```diff
@@ -0,0 +1,264 @@
+    }
+   },
+   "source": [
+    "Expected input is a pyspark dataframe with a column containing image as base64 string for input into VLMs\n",
```
I checked, we def use 'local path' e.g. /Volumes/catalog/schema/volume_name

yenlow replied:
Added support for multiple input types: dcm path, image path, image string

dmoore247 replied:
I don't see, pardon the pun, a path through the `_transform` method and `extract_udf` that uses a path to the file.

yenlow replied:
I have examples for the single file run using dicom path or image string. For the df, I only showed using image str. I can update to show both possible inputs for df run

dmoore247 replied:
I see the path now. Thanks!

yenlow commented on dbx/pixels/dicom/__init__.py:
```diff
@@ -8,3 +8,4 @@
 from dbx.pixels.dicom.dicom_presidio_transformer import PresidioTransformer
 from dbx.pixels.dicom.dicom_thumbnail_extractor import DicomThumbnailExtractor
 from dbx.pixels.dicom.dicom_xform_patcher import DicomPatcher
+from dbx.pixels.dicom.dicom_vlm_phi_detector import VLMTransformer, SYSTEM_PROMPT
```
Agree. Moved it as part of initialization instead of setting it as a constant

dmoore247 commented on dbx/pixels/dicom/dicom_vlm_phi_detector.py:
```diff
@@ -0,0 +1,118 @@
+    """
+    def __init__(self, endpoint, system_prompt, 
+                 temperature:float = 0.0, num_output_tokens:int = 200,
+                 inputCol:str = "img_base64", outputCol:str = "response"):
```
inputCol should default to `path`

dmoore247 commented on dbx/pixels/dicom/dicom_vlm_phi_detector.py:
```diff
@@ -0,0 +1,118 @@
+    """
+    Transformer class to detect PHI in DICOM image using VLM
+    """
+    def __init__(self, endpoint, system_prompt, 
```
This line should have `system_prompt=SYSTEM_PROMPT` so the default system prompt is the one you recommend. This way user does not have to import SYSTEM_PROMPT and then pass it back to this class.

yenlow replied:
agree. I think this code has been superceded

dmoore247 commented on dbx/pixels/dicom/dicom_vlm_phi_detector.py:
```diff
@@ -0,0 +1,118 @@
+        @pandas_udf(
+            "content array<string>, completion_tokens int, prompt_tokens int, total_tokens int, error string"
+        )
+        def extract_udf(imgs_base64: Iterator[pd.Series]) -> Iterator[pd.DataFrame]:
```
This could be up to 277MB plus (`When using uuencode to encode binary data, such as images, you can expect the size of the encoded file to increase by approximately 33% to 40%`) per our data set statistics. 

Would like to see the input be a path to avoid putting images onto the JVM heap.

yenlow replied:
Agree. This has already been changed to paths. You could be looking at the older version of the code

dmoore247 reviewed (changes requested):
Comments inline.
Thanks so much for your hard work on this. Very close.

yenlow commented on dbx/pixels/dicom/dicom_utils.py:
```diff
@@ -113,3 +121,129 @@ def anonymize_metadata(
+        if path.startswith("dbfs:/Volumes"):
+            path = path.replace("dbfs:", "")
+    except Exception as e:
+        logger.error(f"Exception error: {e}. Path may not be a string")
```
Gd idea! Updated the code as such

yenlow commented:
1. Added another example to show _transform with path input as requested
2. Updated NB, README with more setup details
3. Merged in latest changes from main. Please ignore those files for this PR.

dmoore247 commented on dbx/pixels/dicom/dicom_vlm_phi_detector.py:
```diff
@@ -0,0 +1,118 @@
+            )
+
+        except Exception as e:
+            return None, 0, 0, 0, str(e)
```
When raising an error, when logging an error; unless it is PHI/PII, or sensitive, please log and report out the context **values** of the error. E.g. endpoint, input_type, system_prompt, temperature, num_output_tokents (**path**).

dmoore247 commented on notebooks/DE-ID/DICOM - VLMTransformer.ipynb:
```diff
@@ -0,0 +1,752 @@
+   "outputs": [],
+   "source": [
+    "#%pip install openai==1.35.3 pydicom==2.4.4 easyocr==1.7.2 numpy==1.26.4\n",
+    "%pip install -e /Workspace/Users/yen.low@databricks.com/pixel/pixels\n",
```
@yenlow We can't have pip installs from personal repos. cc: @erinaldidb

yenlow replied:
Looks like @erinaldidb already created a config/setup so will suggest running that instead (essentially a sys.path append)

dmoore247 commented on notebooks/DE-ID/DICOM - VLMTransformer.ipynb:
```diff
@@ -0,0 +1,752 @@
+    "\n",
+    "# For .dcm path input (inputCol=\"path\", input_type=\"dicom\")\n",
+    "vlm_transformer = VLMTransformer(endpoint=\"databricks-claude-3-7-sonnet\", \n",
+    "                                 temperature = 0.0,\n",
```
In the first example, pls show the simplest configuration where all defaults are used. 

I think you've provided smart defaults in the VLMTransformer, to include them in the 1st example takes away from the simplification you've performed for the end user. To that end, remove `temperature`, `num_output_tokens`, `inputCol`, `outputCol`, `max_width`,...;

Your 2nd example, where the story is "powerful," do include the overrides the end user is likely to be interested in.

yenlow replied:
Removed default params from first example

dmoore247 commented on dbx/pixels/dicom/dicom_vlm_phi_detector.py:
```diff
@@ -0,0 +1,203 @@
+        # if dicom path, convert to image then base64 string
+        if input_type == "dicom":
+            try:
+                image_base64 = dicom_to_image(path, max_width=768, return_type="str")
```
This doesn't use the parameter from 20 lines above.

dmoore247 commented on dbx/pixels/dicom/dicom_vlm_phi_detector.py:
```diff
@@ -0,0 +1,203 @@
+                for path in batch:
+                    try:
+                        response = extractor.extract(
+                            path, input_type=self.input_type, max_width=self.max_width
```
This looks good here.

erinaldidb reviewed (commented):
We shouldn't merge any PR if at least AWS Integration test is not green
